jetstream/engine/aot_utils.py

```diff
@@ -0,0 +1,238 @@
+"""AOT compilation utils."""
+
+import jax
+import concurrent.futures
+from typing import Any, Optional
+import logging
+from jetstream.engine import engine_api, token_utils
+
+
+def layout_params_and_compile_executables(
+    prefill_engines: Optional[list[engine_api.Engine]] = None,
+    generate_engines: Optional[list[engine_api.Engine]] = None,
+    prefill_params: Optional[list[Any]] = None,
+    generate_params: Optional[list[Any]] = None,
+) -> bool:
+  """Organizes the engines and executables.
+
+  Args:
+      prefill_engines: Prefill only engines.
+      generate_engines: Generate only engines.
+      prefill_params: Prefill only params.
+      generate_params: Generate only params.
+
+  Returns:
+      bool:
+  """
+  prefill_engines = prefill_engines if prefill_engines else []
+  generate_engines = generate_engines if generate_engines else []
+  prefill_params = prefill_params if prefill_params else []
+  generate_params = generate_params if generate_params else []
+
+  compiled_prefills = []
+  compiled_inserts_generate = []
+
+  for i, pe in enumerate(prefill_engines):
+    prefill_compiled = initialize_prefill_jit_cache(
+        prefill_engine=pe,
+        prefill_params=prefill_params[i],
+        prefill_idx=i,
+    )
+    compiled_prefills.append(prefill_compiled)
+
+  for i, ge in enumerate(generate_engines):
+    insert_compiled, generate_compiled = initialize_insert_generate_jit_cache(
+        generate_engine=ge,
+        generate_params=generate_params[i],
+        generate_idx=i,
+    )
+    compiled_inserts_generate.append([insert_compiled, generate_compiled])
+
+  if compiled_prefills and compiled_inserts_generate:
+    return True
+  return False
+
+
+def initialize_prefill_jit_cache(
+    *,
+    prefill_engine: engine_api.Engine,
+    prefill_params: Any,
+    prefill_idx: int,
+):
+  """Precompile all prefill functions in parallel.
+  If we don't do this, then when a new request triggers a new prefill bucket it
+  will take a very long time for that query to come back.
+
+  Args:
+      prefill_engine: A prefill engine to be compiled for.
+      prefill_params: The associated prefill parameters.
+      prefill_idx: Which prefill engine it is.
+  """
+  prefill_buckets = token_utils.DEFAULT_PREFILL_BUCKETS
+  prefill_buckets = [
+      bucket
+      for bucket in prefill_buckets
+      if bucket <= prefill_engine.max_prefill_length
+  ]
+  prefill_engine.prefill_buckets = prefill_buckets
+  if prefill_engine.max_prefill_length not in prefill_buckets:
+    prefill_buckets.append(prefill_engine.max_prefill_length)
+
+  def compile_prefill(length):
+    metadata = prefill_engine.get_tokenizer()
+    vocab = token_utils.load_vocab(metadata.path, metadata.extra_ids)
+    padded_tokens, true_length = token_utils.tokenize_and_pad(
```

Comment:
The shape of padded tokens should be of [batch size x seq len]. So you can do `jnp.ones((batch size, length), dtype='int32')` (or maybe int64 please verify first) instead

Reply:
Yes, int32 is correct. However when using batch_size in the jnp.ones, there is an error with mismatch shape later on. Using without batch size seems to work. Does using batch size perhaps help with the performance degradation issues? Since using without batch size works for now, closing and we can address this in a future PR

```
File "/jetstream-pytorch/jetstream_pt/layers.py", line 365, in reshape_for_broadcast
    assert freqs_cis.shape == (
AssertionError: freqs_cis: (1, 256, 4), x: (1, 256, 32, 64)
```